File: 3040-maximum-number-of-operations-with-the-same-score-ii/3040-maximum-number-of-operations-with-the-same-score-ii.cpp
class Solution {
public:
    map<pair<int, int>, int> m;
    int dfs(int i, int j, int score, vector<int>& nums){
        if(m.count({i, j})){
            return m[{i, j}];
        }
        int maxx = 0;
        if(j - i + 1 >= 2){
            if(nums[i] + nums[i + 1] == score){
                maxx = max(maxx, 1 + dfs(i + 2, j, score, nums));
            }
            if(nums[i] + nums[j] == score){
                maxx = max(maxx, 1 + dfs(i + 1, j - 1, score, nums));
            }
            if(nums[j] + nums[j - 1] == score){
                maxx = max(maxx, 1 + dfs(i, j - 2, score, nums));
            }
        }
        return m[{i, j}] = maxx;
    }
    int maxOperations(vector<int>& nums) {
        int n = nums.size();
        return max({
            dfs(0, n - 3, nums[n - 1] + nums[n - 2], nums),
            dfs(1, n - 2, nums[0] + nums[n - 1], nums),
            dfs(2, n - 1, nums[0] + nums[1], nums)
        }) + 1;
    }
};

File: 3040-maximum-number-of-operations-with-the-same-score-ii/3040-maximum-number-of-operations-with-the-same-score-ii.py
class Solution:
    def maxOperations(self, nums: List[int]) -> int:
        @cache
        def dfs(i, j, s):
            maxx = 0
            if j - i + 1 >= 2:
                if nums[i] + nums[i + 1] == s:
                    maxx = max(maxx, 1 + dfs(i + 2, j, s))
                if nums[i] + nums[j] == s:
                    maxx = max(maxx, 1 + dfs(i + 1, j - 1, s))
                if nums[j] + nums[j - 1] == s:
                    maxx = max(maxx, 1 + dfs(i, j - 2, s))
            return maxx
                
        
        return max(
            dfs(0, len(nums) - 3, nums[-1] + nums[-2]),
            dfs(1, len(nums) - 2, nums[0] + nums[-1]),
            dfs(2, len(nums) - 1, nums[1] + nums[0]),
        ) + 1